pkg/resources/grant_application_role.go

```
@@ -79,54 +86,91 @@ func GrantApplicationRole() *schema.Resource {
-	objectType := parts[1]
+	objectTypeString := parts[1]
+	if objectTypeString == "ACCOUNT_ROLE" {
+		objectTypeString = "ROLE"
+	}
+
+	// first check if either the target account role or application exists
 	targetIdentifier := parts[2]
-	ctx := context.Background()
+	objectType := sdk.ObjectType(objectTypeString)
+	switch objectType {
+	case sdk.ObjectTypeRole:
+		{
+			if _, err := client.Roles.ShowByID(ctx, sdk.NewShowByIdRoleRequest(sdk.NewAccountObjectIdentifierFromFullyQualifiedName(targetIdentifier))); err != nil && errors.Is(err, sdk.ErrObjectNotExistOrAuthorized) {
+				d.SetId("")
+				return diag.Diagnostics{
+					diag.Diagnostic{
+						Severity: diag.Warning,
+						Summary:  "Failed to retrieve account role. Marking the resource as removed.",
+						Detail:   fmt.Sprintf("Id: %s", d.Id()),
+					},
+				}
+			}
+		}
+	case sdk.ObjectTypeApplication:
+		if _, err := client.Applications.ShowByID(ctx, sdk.NewAccountObjectIdentifierFromFullyQualifiedName(targetIdentifier)); err != nil && errors.Is(err, sdk.ErrObjectNotExistOrAuthorized) {
+			d.SetId("")
+			return diag.Diagnostics{
+				diag.Diagnostic{
+					Severity: diag.Warning,
+					Summary:  "Failed to retrieve application. Marking the resource as removed.",
+					Detail:   fmt.Sprintf("Id: %s", d.Id()),
+				},
+			}
+		}
+	}
+	// then check if application role exists
 	grants, err := client.Grants.Show(ctx, &sdk.ShowGrantOptions{
 		Of: &sdk.ShowGrantsOf{
 			ApplicationRole: applicationRoleIdentifier,
 		},
 	})
-	if err != nil {
-		log.Printf("[DEBUG] application role (%s) not found", applicationRoleIdentifier.FullyQualifiedName())
+	if err != nil && errors.Is(err, sdk.ErrObjectNotExistOrAuthorized) {
 		d.SetId("")
-		return nil
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  "Failed to retrieve application role. Marking the resource as removed.",
+				Detail:   fmt.Sprintf("Id: %s", d.Id()),
+			},
+		}
 	}
```

shouldn't we also check the err even if it's not the `ErrObjectNotExistOrAuthorized`? Right now, we discard the error, but maybe in that case we should return an error with `err` as additional context.

---

Fixed.